src/OpenTelemetry.ClrProfiler.Managed/Instrumentation.cs

```diff
@@ -100,6 +105,23 @@ public static void Initialize()
             }
         }
 
+        private static void OnExit(object sender, EventArgs e)
+        {
+            if (Interlocked.Exchange(ref _isExiting, value: 1) != 0)
+            {
+                // OnExit() was already called before
+                return;
+            }
+
+            if (_tracerProvider is not null)
+            {
+                _tracerProvider.Dispose();
+                _tracerProvider = null;
+
+                Log("OpenTelemetry tracer exit.");
+            }
```

Comment:
the code looks as if it could be accessed from multiple threads. if this is so then access to `_tracerProvider` is not synchronized at all

proposals:
1. check if it is not the main thread that is always calling `Initialize` and `OnExit` - if so then maybe we can remove the `Interlocked.Exchange()` and note in comments that the class is accessed always from the same thread, therefore, it does not need to be thread-safe
2. add simple `lock` around `_tracerProvider` usage. it should also help in a situation when `OnExit()` is called before `Initialize()` has set the `_tracerProvider` value. In practice, it should not be a problem but I think it is always better to have thread-safe code. I dislike when the code is "partially thread-safe".

Reply:
maybe it will be easier to go for "proposal 2" and make an issue to check if "proposal 1" is valid

Reply:
Currently OnExit is not subscribed before _traceProvider is initialized.

Reply:
you are right. then `if (_tracerProvider is not null)` can be removed

Reply:
true, gives wrong impression.